src/storage/ExprVisitorBase.cpp

```diff
@@ -0,0 +1,86 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+#include "storage/ExprVisitorBase.h"
+namespace nebula {
+namespace storage {
+void ExprVisitorBase::visit(ConstantExpression *expr) { UNUSED(expr); }
+void ExprVisitorBase::visit(UnaryExpression *expr) { expr->operand()->accept(this); }
+void ExprVisitorBase::visit(TypeCastingExpression *expr) { expr->operand()->accept(this); }
+void ExprVisitorBase::visit(LabelExpression *expr) { fatal(expr); }
+void ExprVisitorBase::visit(LabelAttributeExpression *expr) { fatal(expr); }
+// binary expression
+void ExprVisitorBase::visit(ArithmeticExpression *expr) {
+  expr->left()->accept(this);
+  expr->right()->accept(this);
+}
+void ExprVisitorBase::visit(RelationalExpression *expr) {
+  expr->left()->accept(this);
+  expr->right()->accept(this);
+}
+void ExprVisitorBase::visit(SubscriptExpression *expr) {
+  expr->left()->accept(this);
+  expr->right()->accept(this);
+}
+void ExprVisitorBase::visit(AttributeExpression *expr) { fatal(expr); }
+void ExprVisitorBase::visit(LogicalExpression *expr) {
+  for (auto operand : expr->operands()) {
+    operand->accept(this);
+  }
+}
+// function call
+void ExprVisitorBase::visit(FunctionCallExpression *expr) {
+  for (auto arg : expr->args()->args()) {
+    arg->accept(this);
+  }
+}
+void ExprVisitorBase::visit(AggregateExpression *expr) { fatal(expr); }
+void ExprVisitorBase::visit(UUIDExpression *expr) { UNUSED(expr); }
+// variable expression
+void ExprVisitorBase::visit(VariableExpression *expr) { fatal(expr); }
```

Comment:
Why do we need to define to storage layer? I don't approve of it.

Reply:
The visitor defined in the graph is more like an interface rather than an abstract class, it lacks some default behavior of recursively visiting the child nodes of expression. In addition, for some expression types, storage is impossible to handle, it should be replaced in graph. However, if we need to define multiple visitors, this part of the code needs to be written many times, and it is not possible to determine which ones must be processed by storage and which ones must not be required. Define a base visitor here to handle all the above content. When storage needs a visitor, it is good to directly derive a subclass from the base visitor (except Selection, strictly speaking, Projection, Sort, Limit, etc. also need visitor, because they The object to be processed may also be an expression rather than a simple value)

Reply:
> The visitor defined in the graph is more like an interface rather than an abstract class, it lacks some default behavior of recursively visiting the child nodes of expression. In addition, for some expression types, storage is impossible to handle, it should be replaced in graph. However, if we need to define multiple visitors, this part of the code needs to be written many times, and it is not possible to determine which ones must be processed by storage and which ones must not be required. Define a base visitor here to handle all the above content. When storage needs a visitor, it is good to directly derive a subclass from the base visitor (except Selection, strictly speaking, Projection, Sort, Limit, etc. also need visitor, because they The object to be processed may also be an expression rather than a simple value)

I see, WDYT @critical27 ?